File: benchmarks/poly/factorial/factorial.pl
use Time::HiRes qw(clock_gettime CLOCK_MONOTONIC);

sub fact {
  my ($n) = @_;
  return 1 if $n < 2;
  return $n * fact($n - 1);
}

sub main {
  my $_t0 = clock_gettime(CLOCK_MONOTONIC);
  for my $i (0 .. 9) {
    print fact($i) . "\n";
  }
  printf STDERR "bench_ns=%.0f\n", (clock_gettime(CLOCK_MONOTONIC) - $_t0) * 1e9;
}

main();

File: benchmarks/poly/factorial/factorial.rb
def fact(n)
  return 1 if n < 2
  n * fact(n - 1)
end

def main
  _t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  (0..9).each do |i|
    puts fact(i)
  end
  $stderr.puts "bench_ns=#{Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - _t0}"
end

main
